Improve subresources DB query performance by removing a subquery

clemherreman opened the pull request:
<!-- Please update this template with something that matches your PR -->
| Q             | A
| ------------- | ---
| Bug fix?      | yes (kinda) <!-- if yes, please use the branch of the current version of API Platform -->
| New feature?  | no <!-- if yes, please use the master branch -->
| BC breaks?    | no
| Deprecations? | no
| Tickets       | none
| License       | MIT
| Doc PR        | N/A

## Abstract 

This PR try to change DQL queries generated to fetch subresource, to improve performance on large tables.

## Before

The generated SQL to query a subresource uses a `WHERE ... IN ( <subquery> )`, in order to be recursive when subresources also have subresources (introduced by #1608).

The subquery, on large table (at work 4M+ lines), takes a significant amount of time (2-5s), even with the proper SQL indexes (at least on MySQL).

```SQL
SELECT
    DISTINCT c0_.idpurchase AS idpurchase_0,
    c0_.idpurchase AS idpurchase_1
FROM purchase c0_
LEFT JOIN customer c1_ ON c0_.idcustomer = c1_.idcustomer
-- Some other LEFT JOINs
WHERE c0_.idcustomer IN 
(  -- here please notice the usage of a subquery
    SELECT c16_.idcustomer
    FROM customer c16_
    WHERE c16_.idcustomer = :identifier
)
```

## After

This PRs tries to optimize for the common case, having only one level of subresources, by replacing the subquery via a direct `WHERE <identifier> = <identifier-alias>` **only on the last recursion level**. This means it shouldn't break subresources that have subresources, while still improving the query for the nominal cases.

```SQL
SELECT
    DISTINCT c0_.idpurchase AS idpurchase_0,
    c0_.idpurchase AS idpurchase_1
FROM purchase c0_
LEFT JOIN customer c1_ ON c0_.idcustomer = c1_.idcustomer
-- Some other LEFT JOINs
WHERE c0_.idcustomer = :identifier

```

**This significantly reduces the SQL query time, going from 2.45s to 15ms on my e-commerce app.**

## BC & missing things.

I broke the unit tests, mainly the mocking, but also some tests fails as they compare the previously generated query, with the new one (which makes sense). All the Behat tests are green locally though.

I feel out of my depth here, and if someone would help me writing those, ensuring no BC is done, is would be great.

On my side, I run our rather large behat test suite against my fork, and not tests complained about wrong results, so I am fairly confident I haven't introduced some BC.

# TODO

- [x] Fix Behat tests
- [x] Fix PHPUnit tests

dkarlovi commented on src/Bridge/Doctrine/Orm/SubresourceDataProvider.php:
```diff
@@ -142,7 +142,9 @@ private function buildQuery(array $identifiers, array $context, QueryNameGenerat
         if (!$classMetadata instanceof ClassMetadataInfo) {
-            throw new RuntimeException("The class metadata for $identifierResourceClass must be an instance of ClassMetadataInfo.");
+            throw new RuntimeException(
+                "The class metadata for $identifierResourceClass must be an instance of ClassMetadataInfo."
```
Code style change? I'd revert it.

clemherreman replied:
Reverted. I guess it sneaked out because PHPStorm was too enthusiastic.

dkarlovi commented on src/Bridge/Doctrine/Orm/SubresourceDataProvider.php:
```diff
@@ -198,16 +200,42 @@ private function buildQuery(array $identifiers, array $context, QueryNameGenerat
 
+        $lastIdentifer = $remainingIdentifiers === 1;
+        $association = $classMetadata->hasAssociation($previousAssociationProperty) ? $classMetadata->getAssociationMapping($previousAssociationProperty) : [];
+        $optimizable = $lastIdentifer && (
```
Better explain *why* it's optimizable and make that the variable name here.

clemherreman replied:
I added a long comment about the optimization, when using `$optimizable`. I could not find a variable name that was both concise and explanatory though.

Is that ok with you?

teohhanhui replied:
`$needsSubquery`? (So it'd have the opposite value.)

teohhanhui replied:
Or `$isLeaf` to make it merely descriptive and decoupled from the implementation.

clemherreman replied:
I reworked the way the code is written, to have more readable code branches and variable name. WDYT?

teohhanhui replied:
`$eligibleOneToOneOptimization` -> `$oneToOneBidirectional`
`$eligibleOneToManyOptimization` -> `$oneToManyBidirectional`

Same idea... The names can be merely descriptive.

clemherreman replied:
Done :)

soyuka commented:
> All the Behat tests are green locally though.

That's good, I've covered the edge cases of this in the behat test suite. The queries you show give the same results, therefore no BC break will outcome of this change. You can change the broken phpunit tests by fixing the relevant queries.

teohhanhui commented:
Thanks for doing this! :tada:

clemherreman commented:
Fixed the PHPUnit tests, and rebased against latest `2.5`.

If the comment I added after @dkarlovi feedback is enough, I'd say this is ready to be merged.

teohhanhui commented:
> Behat tests are still failing, but because of either ccode coverage generation failing (?) or call to Dummy::hasRole() failing in the serialization. Both errors seems unrelated to that pull request, isn't it?

Yeah, those are unrelated. Please ignore those failures. :smile:

clemherreman commented:
@teohhanhui Behat tests already covers the correctness of the query results (thanks to @alanpoulain)

Is there a point in checking the generated DQL itself? I doubt so. This is a perf improvement on the query level.

alanpoulain commented on src/Bridge/Doctrine/Orm/SubresourceDataProvider.php:
```diff
@@ -198,11 +198,30 @@ private function buildQuery(array $identifiers, array $context, QueryNameGenerat
-            $qb->andWhere("$alias.$key = :$placeholder");
             $topQueryBuilder->setParameter($placeholder, $value, (string) $classMetadata->getTypeOfField($key));
+
+            // Optimizaion: add where clause for identifiers, but not via a WHERE ... IN ( ...subquery... ).
```
```suggestion
            // Optimization: add where clause for identifiers, but not via a WHERE ... IN ( ...subquery... ).
```

teohhanhui commented on src/Bridge/Doctrine/Orm/SubresourceDataProvider.php:
```diff
@@ -198,11 +198,30 @@ private function buildQuery(array $identifiers, array $context, QueryNameGenerat
             $topQueryBuilder->setParameter($placeholder, $value, (string) $classMetadata->getTypeOfField($key));
+
+            // Optimizaion: add where clause for identifiers, but not via a WHERE ... IN ( ...subquery... ).
+            // Instead we use a direct identifier equality clause, to speed thing up when dealing with large tables.
```
```suggestion
            // Instead we use a direct identifier equality clause, to speed things up when dealing with large tables.
```

teohhanhui commented:
Thanks @clemherreman! :tada:

mxmp210 commented:
Please have a look at [#1542](https://github.com/api-platform/api-platform/issues/1542), As it breaks custom identifiers, feel free to review.
